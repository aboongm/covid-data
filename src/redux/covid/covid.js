import Axios from 'axios';

const baseURL = 'https://api.covid19tracking.narrativa.com/api/';
const date = '2022-05-22';
const url = `${baseURL}${date}`;
// constants
const FETCH_DATA = 'covid/covid/FETCH_DATA';

// initial-state
const initialState = {
  datesData: {},
  totalData: {},
};

// actions
export const fetchDataAction = (dates, total) => ({
  type: FETCH_DATA,
  dates,
  total,
});

export const fetchDataApi = () => async (dispatch) => {
  const returnValue = await Axios.get(url);
  const {
    data: { dates, total },
  } = returnValue;
  dispatch(fetchDataAction(dates, total));
};

// reducer
const covidReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_DATA:
      return {
        ...state,
        datesData: action.dates[date].countries,
        totalData: action.total,
      };
    default:
      return state;
  }
};

export default covidReducer;
