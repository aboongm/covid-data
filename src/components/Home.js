import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchDataApi } from '../redux/covid/covid';
import Countries from './Countries';

const Home = () => {
  const [search, setSearch] = useState('');
  const _ = useSelector((state) => state.covidData.datesData);
  const totalData = useSelector((state) => state.covidData.totalData);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchDataApi());
  }, [dispatch]);

  const filteredData = Object.entries(_).filter((item) => Object.keys(item).some((key) => item[key]
    .toString()
    .toLowerCase()
    .includes(search.toLocaleLowerCase())));

  return (
    <>
      <div className="row m-0 p-0 ">
        <div className="col m-0 p-3  d-flex flex-column align-items-center justify-content-center bg-image">
          <p className="text-white text-center text2">
            <span className="text1 text-center">COVID-19 Statistics</span>
            <br />
            Around The World
          </p>
        </div>
        <div className="col bg-color m-0 p-3 ">
          <p className="text-white text2">
            <span className="text1">TOTAL:</span>
            <br />
            {totalData.today_confirmed}
          </p>
          <p className="text-white text2">
            <span className="text1">DATE:</span>
            <br />
            {totalData.date}
          </p>
        </div>
      </div>
      <div className="text-center bg-color1 m-0 pt-2 pb-3">
        <p className="text-white m-0 p-0 text2">SEARCH BY COUNTRY NAME</p>
        <input
          type="text"
          placeholder="Type a country"
          className="mx-0 mt-2 mb-0 ps-3 py-1 rounded-pill text2"
          value={search}
          onChange={(e) => setSearch(e.target.value)}
        />
      </div>
      <div className="m-0 countries-container bg-color row">
        {filteredData.map((data) => (
          <Countries key={data[0]} data={data[1]} />
        ))}
      </div>
    </>
  );
};

export default Home;
