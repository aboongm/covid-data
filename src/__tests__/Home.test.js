import React from 'react';
import { Provider } from 'react-redux';
import { HashRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import Home from '../components/Home';
import store from '../redux/configureStore';

import { fetchDataAction } from '../redux/covid/covid';

describe('Test for Home', () => {
  test('should render', () => {
    const homepage = render(
      <Provider store={store}>
        <HashRouter>
          <Home />
        </HashRouter>
      </Provider>,
    );
    expect(homepage).toMatchSnapshot();
  });

  it('should return an action with type FETCH_DATA', () => {
    const data = {
      dates: {
        '2022-05-22': {
          countries: {
            Afghanistan: {},
            info: {},
          },
        },
        metadata: {},
        total: {},
        updated_at: '2022-05-23 14:54UTC',
      },
    };
    const action = fetchDataAction(data);
    expect(action.type).toBe('covid/covid/FETCH_DATA');
  });
});
