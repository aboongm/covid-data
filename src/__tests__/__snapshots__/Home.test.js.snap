// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Home should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row m-0 p-0 "
      >
        <div
          class="col m-0 p-3  d-flex flex-column align-items-center justify-content-center bg-image"
        >
          <p
            class="text-white text-center"
          >
            <span
              class="h5 text-center"
            >
              COVID-19 Statistics
            </span>
            <br />
            Around The World
          </p>
        </div>
        <div
          class="col bg-color m-0 p-3 "
        >
          <p
            class="text-white"
          >
            <span
              class="h5"
            >
              Total Cases:
            </span>
            <br />
          </p>
          <p
            class="text-white"
          >
            <span
              class="h5"
            >
              Date:
            </span>
            <br />
          </p>
        </div>
      </div>
      <div
        class="text-center bg-color1 m-0 pt-2 pb-3"
      >
        <p
          class="text-white m-0 p-0"
        >
          Search by country name
        </p>
        <input
          class="mx-0 mt-2 mb-0 ps-3 py-1 rounded-pill"
          placeholder="Type a country"
          type="text"
          value=""
        />
      </div>
      <div
        class="m-0 countries-container bg-color row"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="row m-0 p-0 "
    >
      <div
        class="col m-0 p-3  d-flex flex-column align-items-center justify-content-center bg-image"
      >
        <p
          class="text-white text-center"
        >
          <span
            class="h5 text-center"
          >
            COVID-19 Statistics
          </span>
          <br />
          Around The World
        </p>
      </div>
      <div
        class="col bg-color m-0 p-3 "
      >
        <p
          class="text-white"
        >
          <span
            class="h5"
          >
            Total Cases:
          </span>
          <br />
        </p>
        <p
          class="text-white"
        >
          <span
            class="h5"
          >
            Date:
          </span>
          <br />
        </p>
      </div>
    </div>
    <div
      class="text-center bg-color1 m-0 pt-2 pb-3"
    >
      <p
        class="text-white m-0 p-0"
      >
        Search by country name
      </p>
      <input
        class="mx-0 mt-2 mb-0 ps-3 py-1 rounded-pill"
        placeholder="Type a country"
        type="text"
        value=""
      />
    </div>
    <div
      class="m-0 countries-container bg-color row"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
